### ISSUE TITLE
CParserError: Exception occurred in remote worker.

### ISSUE THREAD

**Issue by blitzd** [blaze/dask#601]
I'm having issues using dask dataframes to analyze any data. It initially appears to work with small sets of the data, but then when anything is run on the dataframe as a whole it throws a CParserError.

```
import dask.dataframe as dd
df = dd.read_csv(r'data.csv', encoding='utf-16', dtype=str)
df.head() # this shows data as expected
df.head().Area.value_counts() # this shows data as expected
df.head(100000).Area.value_counts() # and this shows data as expected
df.Area.value_counts().compute() # this throws the error below
```

I'm running on:

dask/0.7.0/np19py34_0

Error:

```
---------------------------------------------------------------------------
CParserError                              Traceback (most recent call last)
<ipython-input-30-b212bd6245ed> in <module>()
----> 1 t.count().head(5)

C:\Anaconda3\lib\site-packages\dask\dataframe\core.py in head(self, n, compute)
    229 
    230         if compute:
--> 231             result = result.compute()
    232         return result
    233 

C:\Anaconda3\lib\site-packages\dask\base.py in compute(self, **kwargs)
     31 
     32     def compute(self, **kwargs):
---> 33         return compute(self, **kwargs)[0]
     34 
     35     @classmethod

C:\Anaconda3\lib\site-packages\dask\base.py in compute(*args, **kwargs)
     65                 for opt, val in groups.items()])
     66     keys = [arg._keys() for arg in args]
---> 67     results = get(dsk, keys, **kwargs)
     68     return tuple(a._finalize(a, r) for a, r in zip(args, results))
     69 

C:\Anaconda3\lib\site-packages\dask\threaded.py in get(dsk, result, cache, num_workers, **kwargs)
     55     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     56                         cache=cache, queue=queue, get_id=_thread_get_id,
---> 57                         **kwargs)
     58 
     59     return results

C:\Anaconda3\lib\site-packages\dask\async.py in get_async(apply_async, num_workers, dsk, result, cache, queue, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, **kwargs)
    498                 "    dataset.compute(rerun_exceptions_locally=True)\n\n"
    499                 "The original exception and traceback follow below:\n\n"
--> 500                     + str(res) + "\n\nTraceback:\n" + tb)
    501         state['cache'][key] = res
    502         finish_task(dsk, key, state, results, keyorder.get)

CParserError: Exception occurred in remote worker.

Something you've asked dask to compute raised an exception.
That exception and the traceback are copied below.
To use pdb, rerun the computation with the keyword argument
    dask.set_options(rerun_exceptions_locally=True)
    or
    dataset.compute(rerun_exceptions_locally=True)

The original exception and traceback follow below:

Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.

Traceback:
  File "C:\Anaconda3\lib\site-packages\dask\async.py", line 266, in execute_task
    result = _execute_task(task, data)
  File "C:\Anaconda3\lib\site-packages\dask\async.py", line 249, in _execute_task
    return func(*args2)
  File "C:\Anaconda3\lib\site-packages\pandas\io\parsers.py", line 474, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "C:\Anaconda3\lib\site-packages\pandas\io\parsers.py", line 250, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "C:\Anaconda3\lib\site-packages\pandas\io\parsers.py", line 566, in __init__
    self._make_engine(self.engine)
  File "C:\Anaconda3\lib\site-packages\pandas\io\parsers.py", line 705, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "C:\Anaconda3\lib\site-packages\pandas\io\parsers.py", line 1072, in __init__
    self._reader = _parser.TextReader(src, **kwds)
  File "pandas\parser.pyx", line 509, in pandas.parser.TextReader.__cinit__ (pandas\parser.c:4732)
  File "pandas\parser.pyx", line 708, in pandas.parser.TextReader._get_header (pandas\parser.c:7023)
  File "pandas\parser.pyx", line 831, in pandas.parser.TextReader._tokenize_rows (pandas\parser.c:8275)
  File "pandas\parser.pyx", line 1742, in pandas.parser.raise_parser_error (pandas\parser.c:20691)
```

I have also tried changing the engine for the dataframe to 'python' as suggested, but that just results in other issues. To be clear though, pandas reads this dataframe and performs calculations on it as a whole with no issues.

**Comment by mrocklin** [comment]
The `dd.read_csv` function does not read the entire file when determining dtypes, only the first 1000 lines or so.  I suspect that some row more deeply in is causing trouble.  Since you have the ability to load the entire dataframe into memory you could actually see what the full dtypes are.  Do the following differ?

```python
>>> dd.read_csv(r'data.csv', encoding='utf-16', dtype=str).dtypes
>>> pd.read_csv(r'data.csv', encoding='utf-16', dtype=str).dtypes
```

**Comment by blitzd** [comment]
I'm specifying dtypes as str by default though... the output of both of the above is object for all columns.

**Comment by mrocklin** [comment]
Hrm, yes that is odd.  Is your data publicly sharable?  I wouldn't mind taking a closer look at the problem.

**Comment by blitzd** [comment]
This particular data is not shareable unfortunately, though I have narrowed it down further to utf-16 specifically, and a small repro:

```
ar = pd.Series(range(0, 1000000))
br = ar % 7
cr = br * 3.3
dr = br / 1.9836
test_df = pd.concat([ar, br, cr, dr], axis=1)
test_df.to_csv('test_df.csv', encoding='utf-16')

d = dd.read_csv('test_df.csv', encoding='utf-16', dtype=str)

d.count().compute()
```

I tried varying ranges of rows. It will work up to 400000, but somewhere between that and 410000 it starts to die.

I also did the same tests with utf-8 up to 5 million rows, and there were no issues there.

Hope that helps.

**Comment by mrocklin** [comment]
Excellent.  Taking a look now.  I suspect that the number of rows corresponds to the default size of a partition.  `d.npartitions == 2` for this data.

**Comment by mrocklin** [comment]
@blitzd I've tracked down at least part of the problem to #606 .

**Comment by cowlicks** [comment]
@blitzd Could you try specifying the encoding as `utf-16-be` or `utf-16-le` which ever it is.

You can figure out the encoding by reading the first two bytes of the file and comparing them with the constants in the `codecs` package in the python standard library.

**Comment by blitzd** [comment]
@cowlicks In the repro above the file is being created just a few lines before the read, and still shows the same issue. I did try the below, though I think I am even more confused by the results:

```
ar = pd.Series(range(0, 1000000))
br = ar % 7
cr = br * 3.3
dr = br / 1.9836
test_df = pd.concat([ar, br, cr, dr], axis=1)
test_df.to_csv('test_df.csv', encoding='utf-16-le')

d = dd.read_csv('test_df.csv', encoding='utf-16-le', dtype=str)

d.count().compute()
```

out:

```
0             407521
1             407521
2             407521
3             407521
Unnamed: 0    407523
dtype: int64
```

I'm running on Win-10, which is LE. When I write/read the file as utf-16-be I actually get the correct count. When I write/read the file as just uft-16, I continue to get a CParser error. I have verified that the test file does indeed include all 1,000,000 rows when it's written as utf-16-le.

**Comment by cowlicks** [comment]
#739 should close this, see the tests [here](https://github.com/blaze/dask/pull/739/files#diff-bddfeaebc5b26f731ea3fbf5595b1825R750), which pass.

**Comment by cowlicks** [comment]
#739 is in

**Comment by jcrist** [comment]
Fixed by #739, closing.
